File: go/pkg/basecamp/otel/otel.go
// Package otel provides OpenTelemetry integration for the Basecamp SDK.
//
// It implements the basecamp.Hooks interface to provide distributed tracing
// and metrics for all HTTP operations.
//
// # Usage
//
//	import (
//	    "github.com/basecamp/basecamp-sdk/go/pkg/basecamp"
//	    basecampotel "github.com/basecamp/basecamp-sdk/go/pkg/basecamp/otel"
//	)
//
//	hooks := basecampotel.NewHooks()
//	client := basecamp.NewClient(cfg, tokenProvider, basecamp.WithHooks(hooks))
package otel

import (
	"context"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/trace"

	"github.com/basecamp/basecamp-sdk/go/pkg/basecamp"
)

const (
	// instrumentationName is the name used for the tracer and meter.
	instrumentationName = "github.com/basecamp/basecamp-sdk/go"

	// Semantic convention attributes for Basecamp operations
	attrBasecampService      = "basecamp.service"
	attrBasecampOperation    = "basecamp.operation"
	attrBasecampResourceType = "basecamp.resource_type"
	attrBasecampIsMutation   = "basecamp.is_mutation"
	attrBasecampMethod       = "basecamp.method"
	attrBasecampURL          = "basecamp.url"
	attrBasecampAttempt      = "basecamp.attempt"
	attrBasecampStatus       = "basecamp.status"
	attrBasecampCached       = "basecamp.cached"
	attrHTTPMethod           = "http.method"
	attrHTTPURL              = "http.url"
	attrHTTPStatusCode       = "http.status_code"
)

// Hooks implements basecamp.Hooks using OpenTelemetry for tracing and metrics.
type Hooks struct {
	tracer            trace.Tracer
	meter             metric.Meter
	operationDuration metric.Float64Histogram
	requestDuration   metric.Float64Histogram
	requests          metric.Int64Counter
	retries           metric.Int64Counter
}

// operationSpanKey is the context key for operation spans.
type operationSpanKey struct{}

// Ensure Hooks implements basecamp.Hooks at compile time.
var _ basecamp.Hooks = (*Hooks)(nil)

// Option configures Hooks.
type Option func(*Hooks)

// WithTracerProvider sets a custom TracerProvider.
func WithTracerProvider(tp trace.TracerProvider) Option {
	return func(h *Hooks) {
		h.tracer = tp.Tracer(instrumentationName)
	}
}

// WithMeterProvider sets a custom MeterProvider.
func WithMeterProvider(mp metric.MeterProvider) Option {
	return func(h *Hooks) {
		h.meter = mp.Meter(instrumentationName)
	}
}

// NewHooks creates a new OpenTelemetry-based Hooks implementation.
// Uses the global TracerProvider and MeterProvider by default.
func NewHooks(opts ...Option) *Hooks {
	h := &Hooks{
		tracer: otel.Tracer(instrumentationName),
		meter:  otel.Meter(instrumentationName),
	}

	for _, opt := range opts {
		opt(h)
	}

	// Initialize metrics
	var err error

	h.operationDuration, err = h.meter.Float64Histogram(
		"basecamp.operation.duration",
		metric.WithDescription("Duration of Basecamp SDK operations in seconds"),
		metric.WithUnit("s"),
	)
	if err != nil {
		otel.Handle(err)
	}

	h.requestDuration, err = h.meter.Float64Histogram(
		"basecamp.request.duration",
		metric.WithDescription("Duration of Basecamp HTTP requests in seconds"),
		metric.WithUnit("s"),
	)
	if err != nil {
		otel.Handle(err)
	}

	h.requests, err = h.meter.Int64Counter(
		"basecamp.requests",
		metric.WithDescription("Total number of Basecamp HTTP requests"),
		metric.WithUnit("{request}"),
	)
	if err != nil {
		otel.Handle(err)
	}

	h.retries, err = h.meter.Int64Counter(
		"basecamp.retries",
		metric.WithDescription("Total number of Basecamp request retries"),
		metric.WithUnit("{retry}"),
	)
	if err != nil {
		otel.Handle(err)
	}

	return h
}

// spanKey is the context key for the span.
type spanKey struct{}

// OnOperationStart creates a new span for the semantic SDK operation.
func (h *Hooks) OnOperationStart(ctx context.Context, op basecamp.OperationInfo) context.Context {
	spanName := op.Service + "." + op.Operation
	ctx, span := h.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindClient),
		trace.WithAttributes(
			attribute.String(attrBasecampService, op.Service),
			attribute.String(attrBasecampOperation, op.Operation),
			attribute.String(attrBasecampResourceType, op.ResourceType),
			attribute.Bool(attrBasecampIsMutation, op.IsMutation),
		),
	)

	if op.ResourceID != 0 {
		span.SetAttributes(attribute.Int64("basecamp.resource_id", op.ResourceID))
	}

	return context.WithValue(ctx, operationSpanKey{}, span)
}

// OnOperationEnd records the operation result and ends the span.
func (h *Hooks) OnOperationEnd(ctx context.Context, op basecamp.OperationInfo, err error, duration time.Duration) {
	span, ok := ctx.Value(operationSpanKey{}).(trace.Span)
	if !ok || span == nil {
		return
	}
	defer span.End()

	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
	} else {
		span.SetStatus(codes.Ok, "")
	}

	// Record operation duration metric
	if h.operationDuration != nil {
		h.operationDuration.Record(ctx, duration.Seconds(), metric.WithAttributes(
			attribute.String(attrBasecampService, op.Service),
			attribute.String(attrBasecampOperation, op.Operation),
		))
	}
}

// OnRequestStart creates a new span for the HTTP request.
func (h *Hooks) OnRequestStart(ctx context.Context, info basecamp.RequestInfo) context.Context {
	ctx, span := h.tracer.Start(ctx, "basecamp.request",
		trace.WithSpanKind(trace.SpanKindClient),
		trace.WithAttributes(
			attribute.String(attrHTTPMethod, info.Method),
			attribute.String(attrHTTPURL, info.URL),
			attribute.String(attrBasecampMethod, info.Method),
			attribute.String(attrBasecampURL, info.URL),
			attribute.Int(attrBasecampAttempt, info.Attempt),
		),
	)

	return context.WithValue(ctx, spanKey{}, span)
}

// OnRequestEnd records the request result and ends the span.
func (h *Hooks) OnRequestEnd(ctx context.Context, info basecamp.RequestInfo, result basecamp.RequestResult) {
	span, ok := ctx.Value(spanKey{}).(trace.Span)
	if !ok || span == nil {
		return
	}
	defer span.End()

	// Add result attributes
	attrs := []attribute.KeyValue{
		attribute.Bool(attrBasecampCached, result.FromCache),
	}

	if result.StatusCode > 0 {
		attrs = append(attrs, attribute.Int(attrHTTPStatusCode, result.StatusCode))
		attrs = append(attrs, attribute.Int(attrBasecampStatus, result.StatusCode))
	}

	span.SetAttributes(attrs...)

	// Record error if present
	if result.Error != nil {
		span.RecordError(result.Error)
		span.SetStatus(codes.Error, result.Error.Error())
	} else {
		span.SetStatus(codes.Ok, "")
	}

	// Record metrics
	metricAttrs := metric.WithAttributes(
		attribute.String(attrHTTPMethod, info.Method),
		attribute.Bool(attrBasecampCached, result.FromCache),
	)

	if h.requestDuration != nil {
		h.requestDuration.Record(ctx, result.Duration.Seconds(), metricAttrs)
	}

	if h.requests != nil {
		statusAttr := attribute.Int(attrHTTPStatusCode, result.StatusCode)
		if result.Error != nil && result.StatusCode == 0 {
			statusAttr = attribute.String("error", "connection_failed")
		}
		h.requests.Add(ctx, 1, metric.WithAttributes(
			attribute.String(attrHTTPMethod, info.Method),
			statusAttr,
		))
	}
}

// OnRetry records a retry attempt.
func (h *Hooks) OnRetry(ctx context.Context, info basecamp.RequestInfo, attempt int, err error) {
	span, ok := ctx.Value(spanKey{}).(trace.Span)
	if ok && span != nil {
		span.AddEvent("retry",
			trace.WithAttributes(
				attribute.Int("attempt", attempt),
				attribute.String("error", err.Error()),
			),
		)
	}

	if h.retries != nil {
		h.retries.Add(ctx, 1, metric.WithAttributes(
			attribute.String(attrHTTPMethod, info.Method),
			attribute.Int("attempt", attempt),
		))
	}
}

File: typescript/src/hooks/otel.ts
/**
 * OpenTelemetry hooks for the Basecamp SDK.
 *
 * Provides distributed tracing and metrics for API operations.
 * Requires @opentelemetry/api as an optional peer dependency.
 *
 * @example
 * ```ts
 * import { createBasecampClient, otelHooks } from "@37signals/basecamp";
 * import { trace, metrics } from "@opentelemetry/api";
 *
 * const tracer = trace.getTracer("my-app");
 * const meter = metrics.getMeter("my-app");
 *
 * const client = createBasecampClient({
 *   accountId: "12345",
 *   accessToken: "token",
 *   hooks: otelHooks({ tracer, meter }),
 * });
 *
 * // All SDK operations will now create spans and record metrics
 * const projects = await client.projects.list();
 * ```
 */

import type { BasecampHooks, OperationInfo, OperationResult, RequestInfo, RequestResult } from "../hooks.js";

// OpenTelemetry API types (optional peer dependency)
// We use the actual types from @opentelemetry/api when available
interface OtelSpan {
  setAttribute(key: string, value: string | number | boolean): this;
  setStatus(status: { code: number; message?: string }): this;
  recordException(exception: Error): this;
  end(): void;
}

interface OtelTracer {
  startSpan(name: string, options?: { attributes?: Record<string, string | number | boolean> }): OtelSpan;
}

interface OtelMeter {
  createHistogram(name: string, options?: { description?: string; unit?: string }): OtelHistogram;
  createCounter(name: string, options?: { description?: string }): OtelCounter;
}

interface OtelHistogram {
  record(value: number, attributes?: Record<string, string | number | boolean>): void;
}

interface OtelCounter {
  add(value: number, attributes?: Record<string, string | number | boolean>): void;
}

// Span status codes from OpenTelemetry
const SpanStatusCode = {
  UNSET: 0,
  OK: 1,
  ERROR: 2,
} as const;

/**
 * Options for OpenTelemetry hooks.
 */
export interface OtelHooksOptions {
  /**
   * OpenTelemetry tracer instance.
   * Create with: `trace.getTracer("your-app-name")`
   */
  tracer?: OtelTracer;

  /**
   * OpenTelemetry meter instance for metrics.
   * Create with: `metrics.getMeter("your-app-name")`
   */
  meter?: OtelMeter;

  /**
   * Whether to record request-level spans (in addition to operation spans).
   * Defaults to false for reduced noise.
   */
  recordRequestSpans?: boolean;

  /**
   * Prefix for span names.
   * Defaults to "basecamp"
   */
  spanPrefix?: string;

  /**
   * Prefix for metric names.
   * Defaults to "basecamp"
   */
  metricPrefix?: string;
}

/**
 * Internal state for tracking spans and request timing.
 */
interface OtelState {
  /** Active operation spans by key */
  operationSpans: Map<string, OtelSpan>;
  /** Active request spans by key */
  requestSpans: Map<string, OtelSpan>;
  /** Histogram for operation duration */
  operationDuration?: OtelHistogram;
  /** Histogram for request duration */
  requestDuration?: OtelHistogram;
  /** Counter for operations */
  operationCounter?: OtelCounter;
  /** Counter for errors */
  errorCounter?: OtelCounter;
  /** Counter for retries */
  retryCounter?: OtelCounter;
}

/** Counter for generating unique keys */
let keyCounter = 0;

/**
 * Creates a unique key for an operation.
 */
function operationKey(info: OperationInfo): string {
  return `${info.service}.${info.operation}:${info.resourceId ?? ""}:${++keyCounter}`;
}

/**
 * Creates a unique key for a request.
 */
function requestKey(info: RequestInfo): string {
  return `${info.method}:${info.url}:${info.attempt}:${++keyCounter}`;
}

/**
 * Creates OpenTelemetry hooks for distributed tracing and metrics.
 *
 * This function creates hooks that integrate with OpenTelemetry for:
 * - **Tracing**: Creates spans for SDK operations with relevant attributes
 * - **Metrics**: Records histograms for latency and counters for operations/errors
 *
 * Span attributes include:
 * - `basecamp.service`: Service name (e.g., "Todos", "Projects")
 * - `basecamp.operation`: Operation name (e.g., "List", "Create")
 * - `basecamp.resource_type`: Type of resource being accessed
 * - `basecamp.is_mutation`: Whether the operation modifies data
 * - `basecamp.project_id`: Project ID if applicable
 * - `basecamp.resource_id`: Resource ID if applicable
 *
 * Metrics recorded:
 * - `basecamp.operation.duration`: Histogram of operation durations (ms)
 * - `basecamp.request.duration`: Histogram of HTTP request durations (ms)
 * - `basecamp.operations.total`: Counter of total operations
 * - `basecamp.errors.total`: Counter of errors
 * - `basecamp.retries.total`: Counter of retry attempts
 *
 * @param options - Configuration options
 * @returns Hooks object for use with createBasecampClient
 *
 * @example
 * ```ts
 * import { createBasecampClient, otelHooks } from "@37signals/basecamp";
 * import { trace, metrics } from "@opentelemetry/api";
 *
 * // Set up OpenTelemetry first (SDK configuration not shown)
 * const tracer = trace.getTracer("my-app", "1.0.0");
 * const meter = metrics.getMeter("my-app", "1.0.0");
 *
 * const client = createBasecampClient({
 *   accountId: "12345",
 *   accessToken: "token",
 *   hooks: otelHooks({
 *     tracer,
 *     meter,
 *     spanPrefix: "basecamp",
 *     recordRequestSpans: true,
 *   }),
 * });
 *
 * // Operations will create spans and record metrics
 * const todos = await client.todos.list(projectId, todolistId);
 * ```
 */
export function otelHooks(options: OtelHooksOptions = {}): BasecampHooks {
  const {
    tracer,
    meter,
    recordRequestSpans = false,
    spanPrefix = "basecamp",
    metricPrefix = "basecamp",
  } = options;

  // Initialize state
  const state: OtelState = {
    operationSpans: new Map(),
    requestSpans: new Map(),
  };

  // Initialize metrics if meter is provided
  if (meter) {
    state.operationDuration = meter.createHistogram(`${metricPrefix}.operation.duration`, {
      description: "Duration of Basecamp SDK operations",
      unit: "ms",
    });

    state.requestDuration = meter.createHistogram(`${metricPrefix}.request.duration`, {
      description: "Duration of HTTP requests",
      unit: "ms",
    });

    state.operationCounter = meter.createCounter(`${metricPrefix}.operations.total`, {
      description: "Total number of Basecamp SDK operations",
    });

    state.errorCounter = meter.createCounter(`${metricPrefix}.errors.total`, {
      description: "Total number of errors",
    });

    state.retryCounter = meter.createCounter(`${metricPrefix}.retries.total`, {
      description: "Total number of retry attempts",
    });
  }

  return {
    onOperationStart(info: OperationInfo): void {
      // Record operation count (even without tracer)
      state.operationCounter?.add(1, {
        service: info.service,
        operation: info.operation,
        is_mutation: info.isMutation,
      });

      if (!tracer) return;

      const spanName = `${spanPrefix}.${info.service}.${info.operation}`;
      const attributes: Record<string, string | number | boolean> = {
        [`${spanPrefix}.service`]: info.service,
        [`${spanPrefix}.operation`]: info.operation,
        [`${spanPrefix}.resource_type`]: info.resourceType,
        [`${spanPrefix}.is_mutation`]: info.isMutation,
      };

      if (info.projectId) {
        attributes[`${spanPrefix}.project_id`] = info.projectId;
      }
      if (info.resourceId) {
        attributes[`${spanPrefix}.resource_id`] = info.resourceId;
      }

      const span = tracer.startSpan(spanName, { attributes });
      const key = operationKey(info);
      state.operationSpans.set(key, span);
    },

    onOperationEnd(info: OperationInfo, result: OperationResult): void {
      // Find and end the span
      if (tracer) {
        // Find the most recent span for this operation
        for (const [key, span] of state.operationSpans) {
          if (key.startsWith(`${info.service}.${info.operation}:`)) {
            if (result.error) {
              span.setStatus({ code: SpanStatusCode.ERROR, message: result.error.message });
              span.recordException(result.error);
              span.setAttribute("error", true);
              span.setAttribute("error.message", result.error.message);
            } else {
              span.setStatus({ code: SpanStatusCode.OK });
            }
            span.setAttribute("duration_ms", result.durationMs);
            span.end();
            state.operationSpans.delete(key);
            break;
          }
        }
      }

      // Record metrics
      const labels = {
        service: info.service,
        operation: info.operation,
        is_mutation: info.isMutation,
        success: !result.error,
      };

      state.operationDuration?.record(result.durationMs, labels);

      if (result.error) {
        state.errorCounter?.add(1, {
          service: info.service,
          operation: info.operation,
          error_type: result.error.name,
        });
      }
    },

    onRequestStart(info: RequestInfo): void {
      if (!tracer || !recordRequestSpans) return;

      const spanName = `${spanPrefix}.http.${info.method}`;
      const span = tracer.startSpan(spanName, {
        attributes: {
          "http.method": info.method,
          "http.url": info.url,
          "http.attempt": info.attempt,
        },
      });

      const key = requestKey(info);
      state.requestSpans.set(key, span);
    },

    onRequestEnd(info: RequestInfo, result: RequestResult): void {
      // Find and end the span
      if (tracer && recordRequestSpans) {
        for (const [key, span] of state.requestSpans) {
          if (key.startsWith(`${info.method}:${info.url}:${info.attempt}:`)) {
            span.setAttribute("http.status_code", result.statusCode);
            span.setAttribute("http.from_cache", result.fromCache);
            span.setAttribute("duration_ms", result.durationMs);

            if (result.error) {
              span.setStatus({ code: SpanStatusCode.ERROR, message: result.error.message });
              span.recordException(result.error);
            } else if (result.statusCode >= 400) {
              span.setStatus({ code: SpanStatusCode.ERROR, message: `HTTP ${result.statusCode}` });
            } else {
              span.setStatus({ code: SpanStatusCode.OK });
            }

            span.end();
            state.requestSpans.delete(key);
            break;
          }
        }
      }

      // Record request metrics
      state.requestDuration?.record(result.durationMs, {
        method: info.method,
        status_code: result.statusCode,
        from_cache: result.fromCache,
      });
    },

    onRetry(info: RequestInfo, attempt: number, error: Error, _delayMs: number): void {
      state.retryCounter?.add(1, {
        method: info.method,
        attempt,
        error_type: error.name,
      });
    },
  };
}
